sound/soc/intel/boards/sof_sdw_max98373.c

```diff
@@ -55,43 +55,68 @@ static int spk_init(struct snd_soc_pcm_runtime *rtd)
 	return ret;
 }
 
-static int max98373_sdw_trigger(struct snd_pcm_substream *substream, int cmd)
+static int max98373_enable_spk_pin(struct snd_pcm_substream *substream, bool enable)
 {
-	int ret;
-
-	switch (cmd) {
-	case SNDRV_PCM_TRIGGER_START:
-	case SNDRV_PCM_TRIGGER_RESUME:
-	case SNDRV_PCM_TRIGGER_PAUSE_RELEASE:
-		/* enable max98373 first */
-		ret = max_98373_trigger(substream, cmd);
-		if (ret < 0)
-			break;
-
-		ret = sdw_trigger(substream, cmd);
-		break;
-	case SNDRV_PCM_TRIGGER_STOP:
-	case SNDRV_PCM_TRIGGER_SUSPEND:
-	case SNDRV_PCM_TRIGGER_PAUSE_PUSH:
-		ret = sdw_trigger(substream, cmd);
-		if (ret < 0)
-			break;
-
-		ret = max_98373_trigger(substream, cmd);
-		break;
-	default:
-		ret = -EINVAL;
-		break;
+	struct snd_soc_pcm_runtime *rtd = asoc_substream_to_rtd(substream);
+	struct snd_soc_dai *codec_dai;
+	struct snd_soc_dai *cpu_dai;
+	int j;
+	int ret = 0;
+
+	/* set spk pin by playback only */
+	if (substream->stream == SNDRV_PCM_STREAM_CAPTURE)
+		return 0;
+
+	cpu_dai = asoc_rtd_to_cpu(rtd, 0);
+	for_each_rtd_codec_dais(rtd, j, codec_dai) {
+		struct snd_soc_dapm_context *dapm =
+				snd_soc_component_get_dapm(cpu_dai->component);
+		char pin_name[16];
+
+		snprintf(pin_name, ARRAY_SIZE(pin_name), "%s Spk",
+			 codec_dai->component->name_prefix);
+
+		if (enable)
+			ret = snd_soc_dapm_enable_pin(dapm, pin_name);
+		else
+			ret = snd_soc_dapm_disable_pin(dapm, pin_name);
+
+		if (!ret)
+			snd_soc_dapm_sync(dapm);
 	}
 
 	return ret;
 }
 
+static int max98373_sdw_prepare(struct snd_pcm_substream *substream)
+{
+	int ret = 0;
+
+	/* according to soc_pcm_prepare dai link prepare is called first */
+	ret = sdw_prepare(substream);
+	if (ret < 0)
+		return ret;
+
+	return max98373_enable_spk_pin(substream, true);
+}
+
+static int max98373_hw_free(struct snd_pcm_substream *substream)
+{
+	int ret = 0;
+
+	/* according to soc_pcm_hw_free dai link free is called first */
+	ret = sdw_hw_free(substream);
+	if (ret < 0)
+		return ret;
+
+	return max98373_enable_spk_pin(substream, false);
+}
+
```

Comment:
When we boot up and no playback was done the spk widgets are enabled, in theory we should have the same issue what we have w/o this patch after playback, no?

There is no way to get analog loopback with this codec, force handling the pin is OK, but I still have reservation that do we really need to?
What is keeping the codec to power off the playback path when there is no playback running?

Reply:
Can SND_SOC_DAPM_PRE and SND_SOC_DAPM_POST DAPM widget be used? Or the root of the problem is that the DAPM is not running for some reason?

Reply:
> When we boot up and no playback was done the spk widgets are enabled, in theory we should have the same issue what we have w/o this patch after playback, no?

no, please check sof_sdw_mx8373_late_probe in this file. The spk pin is disabled.

> There is no way to get analog loopback with this codec, force handling the pin is OK, but I still have reservation that do we really need to?
> What is keeping the codec to power off the playback path when there is no playback running?

please check https://github.com/thesofproject/linux/commit/a6c4e1aea4502b587297eac6cc44ae89d896c268

Reply:
> > When we boot up and no playback was done the spk widgets are enabled, in theory we should have the same issue what we have w/o this patch after playback, no?
> 
> no, please check sof_sdw_mx8373_late_probe in this file. The spk pin is disabled.

I stand corrected.

> > There is no way to get analog loopback with this codec, force handling the pin is OK, but I still have reservation that do we really need to?
> > What is keeping the codec to power off the playback path when there is no playback running?
> 
> please check [a6c4e1a](https://github.com/thesofproject/linux/commit/a6c4e1aea4502b587297eac6cc44ae89d896c268)

I just don't see the path which is claimed in that commit:
```
(HiFi Playback) -> Amp Enable  -> DAI Sel Mux  (left/right/LeftRight) ->  BE OUT  ->  Spk 

VMON ->                   -> Voltage Sense ->
        VI Sense (switch)                     (HiFi Capture)
IMON ->                   -> Current Sense ->

FBMON -> SpkFB Sense (switch) -> Speaker FB Sense -> (HiFi Capture)
```
I don't see anything loopbacking IV/FB sense to Spk.

Reply:
loopback is supported in max98373 codec hw.

Reply:
but it is not modeled in DAPM, DAPM should be shutting down the playback path after playback stop, no (5 seconds or if you have different dapm shutdown time set)?
The only bit which is handled by the playback path is the `Amp Enable`, which is bit 0 of MAX98373_R202B_PCM_RX_EN and this should be disabled when the playback stops.
Do you have DAPM state captured without this patch after a playback? What is keeping DAPM to not power down?

Reply:
Looking at the MAX98374 data sheet which I can find I would guess that the correct fix is to move the bit 0 of MAX98373_R2043_AMP_EN (SPK_EN) handling to DAPM. It is just left enabled in the max98373_probe() and never touched again.
According to the MAX98374 pdf (page 49), if SPK_EN is 0 then the amp will enter to full shutdown.

Reply:
@ujfalusi even when playback stops, there is a firmware level loop between IV capture and speaker playback, so even when playback starts the codec remains active.

The work-around for this is to play with the pin state to force DAPM to shut down this loop as well and hence the rest of the graph/device.

I will agree it's not super elegant, there's nothing tying IV feedback to speaker output.

Reply:
@plbossart, thank you for the detail, but I still don't see how that is represented in DAPM is this loopback DAPM fragment comes via topology? If the firmware level loop is not visible for Linux then we should be shutting down the codec when it become idle.
It would be nice to see how the DAPM looks like.
Is it so that we have an extra DAPM route added between the IV feedback and speaker output or do we have a widget in between to make the capture to trigger in the firmware?

Or handle the internal loopback along with the playback in firmware?

I just want to understand why DAPM it behaves how it behaves, it might be possible to convert the firmware internal loopback handling into a simple DAMP widget which would enable/disable based on the codec DAPM?

Reply:
See topology below

![sof-tgl-sdw-max98373-rt5682](https://user-images.githubusercontent.com/1149035/122801580-ed636a00-d289-11eb-94fc-acc633f4e819.png)

The loop from ALH259.IN to ALH0x102.OUT does not depend no any PCM activity, so it's considered by DAPM as a closed circuit that needs to remain powered at all times. The only what to prevent this patch from remaining powered if to play with a pin status that disables this loop.

Reply:
Ah, I see. We want to have the VI/FB sense data to be captured to BUF4.1 when the codec is active even if user space does not want to have that data, so we connect the ALH259.IN to ALH0x102.OUT which will bring up the path when complete. Do we really move the captured data from BUF4.1 to BUF3.2 (and then to BUF3.1 and finally to speaker)?

Is it so that the ALH259.IN must be connected to a real speaker? Can it be routed to a virtual output? Like
ALH259.IN -> BUF4.1 -> MUXDEMUX4.0 -> BUF3.2 -> VIRTUAL1.OUT

If this is also possible then it might be more elegant if we have SPK_E for the speaker and from the event handler we enable/disable the virtual output based on PMU/PMD event on the playback. This would make the capture from ALH259.IN to start/stop following the playback activity and also allow normal DAPM operation?

Reply:
@RanderWang, @plbossart, this patch will bring in difference how the driver will behave in case when the stream is stopped but the stream is not closed (I think suspend does that?): before this patch you have forced the speaker to be disabled in trigger:stop, with this patch the speaker remains enabled until the playback stream is closed.

Will this bring in different issue?

Reply:
> Ah, I see. We want to have the VI/FB sense data to be captured to BUF4.1 when the codec is active even if user space does not want to have that data, so we connect the ALH259.IN to ALH0x102.OUT which will bring up the path when complete. Do we really move the captured data from BUF4.1 to BUF3.2 (and then to BUF3.1 and finally to speaker)?

BUF4.1->BUF 3.2 is required since we change the number of channels in the input path after the DEMUX.
BUF3.2 -> BUF 3.1 is also required since we modify the input to generate a new output signal. Also the feedback maybe for 2 speakers.


> Is it so that the ALH259.IN must be connected to a real speaker? Can it be routed to a virtual output? Like
> ALH259.IN -> BUF4.1 -> MUXDEMUX4.0 -> BUF3.2 -> VIRTUAL1.OUT

It's connected to a Source Data port on the amplifier side. It cannot be virtual.

> If this is also possible then it might be more elegant if we have SPK_E for the speaker and from the event handler we enable/disable the virtual output based on PMU/PMD event on the playback. This would make the capture from ALH259.IN to start/stop following the playback activity and also allow "normal" DAPM operation?

Reply:
> @RanderWang, @plbossart, this patch will bring in difference how the driver will behave in case when the stream is stopped but the stream is not closed (I think suspend does that?): before this patch you have forced the speaker to be disabled in trigger:stop, with this patch the speaker remains enabled until the playback stream is closed.
> 
> Will this cause issue?

It doesn't hurt to double check the flow for SUSPEND/RESUME. The rest should be fine.